### PR TITLE
[ci] Test for new_release_request action of project controller

### PR DESCRIPTION
**Pull request by eduardoj**

### CODE REVIEW

**Comment by mdeniz** [comment]
[(3) Write test for app/controllers/webui/project_controller.rb#new_release_request](https://trello.com/c/fRorGYOg/25-3-write-test-for-app-controllers-webui-project-controller-rb-new-release-request)

File: src/api/spec/controllers/webui/project_controller_spec.rb [hunk 1]
```diff
@@ -917,4 +917,64 @@
       it { expect(package.attribs.where(attrib_type: attribute_type)).to be_empty }
     end
   end
+
+  describe 'POST #new_release_request' do
+    let(:project) { create(:maintenance_incident_project) }
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: Could you use 

``
  let(:apache_maintenance_incident_project) { create(:maintenance_incident_project, name: 'ApacheMI') }
``
?

File: src/api/spec/controllers/webui/project_controller_spec.rb [hunk 2]
```diff
@@ -917,4 +917,64 @@
...
+      it { expect(response).to redirect_to( project_show_path(project)) }
+    end
+
+    context 'when raises an APIException' do
+      before do
+        post :new_release_request, params: { project: project }
+      end
+
+      it { expect(flash[:error]).to eq 'Internal problem while release request creation' }
+      it { expect(response).to redirect_to( project_show_path(project)) }
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: Please remove whitespace after (

File: src/api/spec/controllers/webui/project_controller_spec.rb [hunk 3]
```diff
@@ -917,4 +917,64 @@
...
+
+    context 'with skiprequest param' do
+      before do
+        post :new_release_request, params: { project: project, skiprequest: true }
+      end
+
+      it { expect(response).to redirect_to( project_show_path(project)) }
+    end
+
+    context 'when raises an APIException' do
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: Out of curiosity: How do you trigger the APIException here?

File: src/api/spec/controllers/webui/project_controller_spec.rb [hunk 4]
```diff
@@ -917,4 +917,64 @@
...
+
+    context 'when is successful' do
+      before do
+        allow_any_instance_of(BsRequest).to receive(:save!).and_return(true)
+        allow_any_instance_of(BsRequest).to receive(:number).and_return(1)
+        post :new_release_request, params: { project: project }
+      end
+
+      it { expect(flash[:success]).to eq "Created maintenance release request <a href='http://test.host/request/show/1'>1</a>" }
+      it { expect(response).to redirect_to( project_show_path(project)) }
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: Please remove whitespace after (

File: src/api/spec/controllers/webui/project_controller_spec.rb [hunk 5]
```diff
@@ -917,4 +917,64 @@
...
+       BsRequestAction::UnknownTargetPackage].each do |exception_class|
+        it_behaves_like 'a non APIException', exception_class
+      end
+    end
+
+    context 'when is successful' do
+      before do
+        allow_any_instance_of(BsRequest).to receive(:save!).and_return(true)
+        allow_any_instance_of(BsRequest).to receive(:number).and_return(1)
+        post :new_release_request, params: { project: project }
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: I wouldn't mock here as setting the attributes from the params is the main logic in this action. So I would prefer to set the 

``
params[:description]
params[:project]
``
and check if it correctly creates the BsRequest and BsRequestAction

Comment (eduardoj) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: I agree with you that the lines of creating an instance of a `BsRequest` class with an `BsRequestActionMaintenanceRelease` action must be tested. Also there are no more places in the the application where you can see a creation of a `BsRequest` with a `BsRequestActionMaintenanceRelease`.

On the other hand, all the lines of the `new_release_request` acion are covered, and in a strict way, the test of this action of the `project` controller is finished.

Can we agree on closing this pull request, and create other pull request for testing `BsRequest` and `BsRequestActionMaintenanceRelease` models? There are no model tests for them now.

File: src/api/app/controllers/webui/project_controller.rb [hunk 6]
```diff
@@ -358,7 +358,7 @@ def update
     authorize @project, :update?
     respond_to do |format|
       if @project.update(project_params)
-        format.html { redirect_to( project_show_path(@project), notice: 'Project was successfully updated.') }
+        format.html { redirect_to(project_show_path(@project), notice: 'Project was successfully updated.') }
```
Comment (ChrisBr) [inline src/api/app/controllers/webui/project_controller.rb]: Please remove this line because it is not related to your commit.

File: src/api/spec/controllers/webui/project_controller_spec.rb (hunk 3 above)
Comment (eduardoj) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: I just forgot to add this line:

`allow_any_instance_of(BsRequest).to receive(:save!).and_raise(APIException)`

File: src/api/spec/controllers/webui/project_controller_spec.rb [hunk 7]
```diff
@@ -79,7 +79,7 @@
         project.reload
       end
 
-      it { expect(response).to redirect_to( project_show_path(project)) }
+      it { expect(response).to redirect_to(project_show_path(project)) }
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: This line is also unrelated to your commit, so please remove it!

File: src/api/spec/controllers/webui/project_controller_spec.rb [hunk 8]
```diff
@@ -917,4 +917,65 @@
       it { expect(package.attribs.where(attrib_type: attribute_type)).to be_empty }
     end
   end
+
+  describe 'POST #new_release_request' do
+    let(:apache_maintenance_incident_project) { create(:maintenance_incident_project, name: 'ApacheMI') }
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project_controller_spec.rb]: With my comment I meant to drop the whole line because it is already defined in line 14.
